[RA2 Ch06] Multi-tenancy and workload isolation Gap (#1623)
cntt-n/CNTT: doc/ref_arch/kubernetes/chapters/chapter06.md

<a name="6.2.2"></a>
### 6.2.2 Multi-tenancy and workload isolation with Kubernetes

**Related requirements:** `e.man.004` `sec.ci.008` `sec.wl.005``sec.wl.006` `req.inf.ntw.03` 
**Baseline project:** none

> **Gap description:** Today, Kubernetes lacks hard multi-tenancy capabilities that give the ability to allow untrusted tenants to share infrastructure resources. This presents a security problem when operators seek to separate workloads by categorization or even just production vs non-production. Further, tenant networks need to be both segregated, but still centrally administered and maintained. Beyond just security, this also presents an operational problem. Trying to deploy too many CNFs into the same cluster could result in version conflicts, configuration conflicts, and problems with software life cycle management. Finally, without proper isolation there is an increased risk of cascading failures. 

Reviewer: is "untrusted" needed?

What does "this" refer to in: "Beyond just security, this also presents an operational problem."

Since "workload" is used within the second sentence, can we replace "CNFs" by "workloads" in "Trying to deploy too many CNFs into the same cluster ..."

<a name="6.3"></a>
## 6.3 Proposals & Resolution

### 6.3.2 Multi-tenancy and workload isolation with Kubernetes

> Kubernetes is not a single cluster solution. This has been demonstrated across the industry from case studies at prominent companies like [Twitter](https://www.alibabacloud.com/blog/what-can-we-learn-from-twitters-move-to-kubernetes_595156), [USA Today](https://medium.com/usa-today-network/there-and-back-again-scaling-multi-tenant-kubernetes-cluster-s-67afb437716c), [Zalando](https://www.youtube.com/watch?v=LpFApeaGv7A), and [Alibaba](https://www.cncf.io/blog/2019/12/12/demystifying-kubernetes-as-a-service-how-does-alibaba-cloud-manage-10000s-of-kubernetes-clusters/) to the bi-annual CNCF survey that finds that the number of clusters being deployed within an organization is growing. While there are many reasons behind the multi cluster paradigm, examining the gap above we find that a multi cluster solution can address many of these problems like security and software life cycle management. 

Reviewer: ```suggestion
Kubernetes is not a single cluster solution. This has been demonstrated across the industry from case studies at prominent companies like [Twitter](https://www.alibabacloud.com/blog/what-can-we-learn-from-twitters-move-to-kubernetes_595156), [USA Today](https://medium.com/usa-today-network/there-and-back-again-scaling-multi-tenant-kubernetes-cluster-s-67afb437716c), [Zalando](https://www.youtube.com/watch?v=LpFApeaGv7A), and [Alibaba](https://www.cncf.io/blog/2019/12/12/demystifying-kubernetes-as-a-service-how-does-alibaba-cloud-manage-10000s-of-kubernetes-clusters/). The bi-annual CNCF survey finds that the number of clusters being deployed within an organization is growing. While there are many reasons for using multi clusters, a multi cluster solution can address many of the gaps listed above such as security and software life cycle management. 
```

> Without hard multi tenancy within a clusters, separate clusters must be used to provide adequate separation for different types of workloads like production vs. non production or per categorization. Having multiple clusters to deploy CNFs into allows operators to chose similar CNFs together while segregating those with different lifecycles from each other. CNFs deployed into the same cluster can be upgraded together to reduce the operational load while CNFs that require different versions, configurations, and dependencies can run in separate clusters and be upgraded independently.

Reviewer 2: I think in our industry the main question is the execution of CNF-s from different vendors in the same Kubernetes cluster.

Author: is there different wording you would like or something you would like to add?

Reviewer 2: Maybe extend this part "separation for different types of workloads like production vs. non production or per categorization." to something like "separation for different types of workloads based on their vendors, type like production vs. non production or per categorization.".

Author: Thanks just updated it

> Without hard multi tenancy within a clusters, separate clusters must be used to provide adequate separation for different types of workloads based on their vendors, type like production vs. non production, or per categorization. Having multiple clusters to deploy CNFs into allows operators to chose similar CNFs together while segregating those with different lifecycles from each other. CNFs deployed into the same cluster can be upgraded together to reduce the operational load while CNFs that require different versions, configurations, and dependencies can run in separate clusters and be upgraded independently.

Reviewer 3: The wording "hard multi tenancy within a cluster" is unclear. Either you have 1 tenant or more in a cluster.
Do you mean strong isolation requirements between some CNFs? 
We need to explain here first that CNFs can be put into different clusters to achieve better isolation. Cluster nodes can be VMs on the same server or even separated on hardware level.
The reasons to put CNFs into different clusters may be various. 
- isolation for different tenants/consumers of the CNFs 
- isolation to support independent lifecycle.

Second comment: Could you elaborate the last sentence about upgrading? I think you can upgrade CNFs together also in different clusters ;-).

Author: Removed "hard" and added strong isolation
added your reasons
tried to make upgrading as required

Does that address your comments?

Reviewer 3: Yes. 
The grammar is a bit difficult to read, but it is ok. Thanks.

> Without multi tenancy within a clusters, separate clusters must be used to provide adequate separation for CNFs that require strong isolation. Putting CNFs may need to be separated for various reasons including different types of workloads based on their vendors, type like production vs. non production, per categorization, or supporting independent lifecycles. Having multiple clusters to deploy CNFs into allows operators to chose similar CNFs together while segregating those with different lifecycles from each other. CNFs deployed into the same cluster can be upgraded together to reduce the operational load while CNFs that require different versions, configurations, and dependencies can run in separate clusters and be upgraded independently if needed.

Reviewer: ```suggestion
Without multi tenancy within a cluster, separate clusters must be used to provide adequate separation for workloads that require strong isolation. Workloads may need to be separated for various reasons including the nature of workloads, workloads based on their vendors, workload deployment type like production and non-production,  and support for different workload lifecycles. For example, having multiple clusters to deploy workloads allows operators to deploy workloads with similar lifecycles together while segregating those with different lifecycles from each other. Workloads deployed into the same cluster can be upgraded together to reduce the operational load while workloads that require different versions, configurations, and other dependencies can run in separate clusters and be upgraded independently if needed.
```

> If running multiple clusters is the only solution to meeting these workload and infrastructure requirements, the operational burden of this model must also be considered. Running a multitude of clusters at scale could be a massive operational challenge if done manually. Any operator considering running Kubernetes at scale should carefully evaluate their multi cluster management strategy including the management of the applications within those clusters.

Reviewer: ```suggestion
If running multiple clusters is the only solution to meeting these workload and infrastructure requirements, the operational burden of this model must also be considered. Running a multitude of clusters at scale could be a massive operational challenge if done manually. Any operator considering running Kubernetes at scale should carefully evaluate their multi cluster management strategy including the management of the workloads within those clusters.
```
